packages/@aws-cdk/aws-eks/lib/cluster.ts

```diff
@@ -2138,10 +2146,15 @@ class ImportedCluster extends ClusterBase {
     this.awscliLayer = props.awscliLayer;
     this.kubectlMemory = props.kubectlMemory;
     this.clusterHandlerSecurityGroup = props.clusterHandlerSecurityGroupId ? ec2.SecurityGroup.fromSecurityGroupId(this, 'ClusterHandlerSecurityGroup', props.clusterHandlerSecurityGroupId) : undefined;
-    this.kubectlProvider = props.kubectlProvider;
     this.onEventLayer = props.onEventLayer;
     this.prune = props.prune ?? true;
 
+    if (props.kubectlProvider) {
+      this.kubectlProvider = props.kubectlProvider;
+    } else if (props.createKubectlProvider) {
+      this.kubectlProvider = new KubectlProvider(this, 'KubectlProvider', { cluster: this });
+    }
+
```

Comment:
I'm not sure this is necessary. Why not just supply it in `props.kubectlProvider`?

Reply:
@comcalvi kubectlProvider itself has a `cluster` property that will not be defined until after `Cluster.fromClusterAttributes()` is constructed.  To define kubectlProvider with the cluster being imported it must be created in the constructor for the cluster.

I explain this in the issue.  https://github.com/aws/aws-cdk/issues/23409

Reply:
I think normally this would have been written like:
```
this.kubectlProvider = props.kubectlProvider ?? new KubectlProvider(this, 'KubectlProvider', { cluster: this });
```
I made a toggle to avoid changing things too much for existing imported clusters.

Reply:
the solution is to add an `addKubectlProvider()` method, so that `Cluster`s can add a `KubectlProvider` after the Cluster has been constructed. Please remove this prop and add that instead.

Reply:
kubectlProvider is a readonly property.  It can only be modified in the constructor.  This seems to be a cdk convention (I proposed changing this, but was rejected in the past.)

Reply:
If the createKubectlProvider toggle is too unusual, we can go with the more typical but also breaking change of always creating the KubectlProvider when props.kubectlProvider is undefined.

Reply:
It can still be `readonly`, but you'll have to create an `_kubectlProvider` property that is not `readonly`. The `addDefaultCloudMapNamespace()` method implements this exact pattern: https://github.com/aws/aws-cdk/blob/v1.186.1/packages/@aws-cdk/aws-ecs/lib/cluster.ts#L294

Reply:
The problem with taking this approach is that kubectlProvider is already a readonly property.  I don't like the idea of creating a additional `private _kubectlProvider` that if set with a method would cause kubectlProvider to still be undefined.

Reply:
we're looking into a workaround for this issue, since none of the solutions we have right now are ideal.

I'll update this when we decide on a course of action.

Reply:
The friction here is that this sort of operation is not ergonomic and doesn't fit neatly into the API. We came up with this workaround that shares the same kubectl provider between stacks, so you don't have the lambda sprawl:

lib:
```
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as eks from 'aws-cdk-lib/aws-eks';

export class EksClusterProviderStack extends cdk.Stack {
  public readonly cluster: eks.ICluster;
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.cluster = new eks.Cluster(this, 'cluster', {
      version: eks.KubernetesVersion.V1_22,
    });
  }
}

export interface ClusterImportStackProps extends cdk.StackProps {
  cluster: eks.ICluster;
}

export class EksClusterImportStack extends cdk.Stack {
  public readonly importedCluster: eks.ICluster;
  constructor(scope: Construct, id: string, props: ClusterImportStackProps) {
    super(scope, id, props);

    const temp = eks.Cluster.fromClusterAttributes(this, `tempCluster`, {
      clusterName: props.cluster.clusterName,
      kubectlRoleArn: props.cluster.kubectlRole?.roleArn,
    });
    const kubectlProvider = eks.KubectlProvider.getOrCreate(this, temp);

    this.importedCluster = eks.Cluster.fromClusterAttributes(this, 'importedCluster', {
      clusterName: props.cluster.clusterName,
      kubectlProvider
    });
  }
}

export interface HelmStackProps extends cdk.StackProps {
  cluster: eks.ICluster;
  chart: string,
  repository: string,
}

export class HelmStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: HelmStackProps) {
    super(scope, id, props);

    new eks.HelmChart(this, 'helmChart', {
      cluster: props.cluster,
      chart: props.chart,
      repository: props.repository,
    });
  }
}
```

bin:
```import * as cdk from 'aws-cdk-lib';
import { EksClusterProviderStack, EksClusterImportStack, HelmStack } from '../lib/eks-stack';

const app = new cdk.App();

const cluster = new EksClusterProviderStack(app, 'EksClusterProviderStack', { env: { region: 'eu-central-1' }} ).cluster;

const clusterImportStack = new EksClusterImportStack(app, 'EksClusterImportStack', {
  cluster,
});

const importedCluster = clusterImportStack.importedCluster;

const helmStackA = new HelmStack(app, 'HelmStack', {
  cluster: importedCluster,
  chart: 'nginx-ingress',
  repository: 'https://helm.nginx.com/stable',
});

const helmStackB = new HelmStack(app, 'HelmStack2', {
  cluster: importedCluster,
  chart: 'hello-world',
  repository: 'https://helm.github.io/examples',
});

const helmStackC = new HelmStack(app, 'HelmStack3', {
  cluster: importedCluster,
  chart: 'nginx-service-mesh',
  repository: 'https://helm.nginx.com/stable',
});
```

the key is the `EksClusterImportStack`, which adds the `KubectlProvider` to the imported cluster